### DISCUSSION TITLE
got time wrong when sleep at microseconds granularity

### DISCUSSION THREAD

**Discussion by baiwfg2** [bpftrace/bpftrace#3143] [Q&A]
Hi. Here's my simple test program:
```c
#include <stdio.h>
#include <stdlib.h>
#include <sys/unistd.h>

__attribute__ ((noinline)) void t_sleep(int us)
{
    usleep(us);
}

int main(int argc, char const *argv[])
{
    int us = atoi(argv[1]);
    t_sleep(us);
    return 0;
}

```

bpftrace version:v.0.20.3

My bt file:
```
uprobe:./a.out:t_sleep 
{
    @t4[tid] = nsecs;
}

uretprobe:./a.out:t_sleep
{
    printf("time: %ld us\n", (nsecs - @t4[tid]) / 1e3);
}

END
{
    clear(@t4);
}
```

Here's my results for running multiple times:
![image](https://github.com/bpftrace/bpftrace/assets/5157680/e339110d-0293-4481-93d4-2a356c912310)

My finding is that the bigger parameter I give, the more precise bpftrace print timing. If I give a few microseconds, it's no longer very correct. I guess it's related to CPU schedule.

1. Does it mean that try not trace very fast function(microseconds level), since it may cause very noisy data ?
2. Is that expected or need optimization in the future ?

**Comment by jordalgo** [comment]
Out of curiosity, if you add some time calculations in `t_sleep` itself do you see consistent results for different params? I will say that factoring in the scheduler and the added time to go back and forth between kernel and userspace to handle uprobes I'm not surprised that the smaller amount of time shows higher error bounds. I'm also curious about what happens if, instead of a sleep, you add a loop for like 1M items or something.

**Reply by ajor** [reply]
I think you're right - the scheduler is the answer. bpftrace is reporting more-or-less the right numbers and it's `usleep` that isn't sleeping for the expected amount of time.

The [man page of usleep](https://man7.org/linux/man-pages/man3/usleep.3.html) clarifies that it can sleep for more than the requested length of time:
> The usleep() function suspends execution of the calling thread for **(at least)** usec microseconds.

I modified the original program to print out the timestamp before and after the `t_sleep` function and these are the results I got:
```
$ ./a.out 10000                                                                                                                               
10077
$ ./a.out 10000                                                                                                                               
10098
$ ./a.out 10000                                                                                                                               
10100
$ ./a.out 1000                                                                                                                                
1077  
$ ./a.out 1000                                                                                                                                
1080                                                                                                                                                                 
$ ./a.out 1000
1080                                                                                                                                                                 
$ ./a.out 100                                                                                                                                 
179   
$ ./a.out 100                                                                                                                                 
177       
$ ./a.out 100
178
```

Tracing these with bpftrace gave roughly the same numbers:
```
Attaching 2 probes...                                                                                                                                                
time: 10064 us
time: 10080 us                                                                                                                                                       
time: 10080 us
time: 1062 us
time: 1063 us
time: 1063 us
time: 164 us
time: 164 us
time: 163 us
^C
```

**Comment by ajor** [comment]
(see previous comment - I think the time difference comes from `usleep`, not bpftrace)

> Does it mean that try not trace very fast function(microseconds level), since it may cause very noisy data ?

bpftrace does add on a few (normally single-digit microseconds) of latency, so for very short functions it can add a noticeable overhead. The large difference (hundreds of microseconds) you're seeing here seems to come from the kernel's scheduler, not bpftrace though.

> Is that expected or need optimization in the future ?

I don't think there's anything for bpftrace to do here. The big difference you're seeing comes from the scheduler, and the small overhead of tracing comes from the kernel's tracing implementation. There's work actively going on to reduce the uprobe overhead to sub-microsecond levels, but we'll get the benefits automatically in newer kernels.
